Cosimulation feature 1

PandoraQS opened the pull request:
**Co-Simulation Process:**

- Updated MaestroManager
     - Added methods to start and stop Maestro Engine, and run the Co-Simulation.
     - Implemented session management to track ongoing simulations.
     - New status updates for real-time UI feedback.
- Enhanced CosimulationAPI (IPC Communication)
    - Introduced IPC invoke/emit methods for communication between React and Electron.
    - Added new API functions   
- Updated useCosimulation Hook:
    - Added state management for
    - Implemented new event listeners
 
**UI**
- Integrated new UI Components for Co-Simulation:
     - Main/Landing page(Main.tsx)
     - CoSimulation page (CoSimulation.tsx)
          - Displays the current simulation state.
          - Shows the results path after completion.

**Error handling with Snackbar notifications**
- Centralized error handling in ErrorHandler.ts.
- Errors from Maestro or IPC are displayed in a Snackbar notification.

**Documentation**
- Created `User Guide.md`
     - Includede xplanations on how to select a project, start Maestro, run a simulation, and retrieve results.
     - Highlighted the **necessary** project folder structure.
- Updated the Class Diagram to reflect the latest changes in MaestroManager, CosimulationAPI, and other components.
- Refactored Sequence Diagrams
- Updated the Package Diagram with new component structure and logic.

**Bug Fixes & Repository Cleanup**
- Fixed Git corruption issues
- Improved error handling messages inside the application.

PandoraQS committed:
Revert "cleaned up the code from comments and unused code"

This reverts commit e9df490f4316ef56a1c5b09c35f1dbc8a0277ceb.

PandoraQS committed:
Merge branch 'development' of https://github.com/PandoraQS/into-cps-application into development

PandoraQS committed:
fixed serve port

PandoraQS committed:
feature 1 - cosimulation page and coe server

prasadtalasila commented on src/components/Cosimulation/Cosimulation.tsx:
```diff
@@ -0,0 +1,51 @@
+        console.log('readJsonFile exists:', typeof window?.electronAPI?.readJsonFile === 'function');
+  
+        console.log('Fetching JSON from:', filePath);
+        const data = await window?.electronAPI?.readJsonFile(filePath); // Dovrebbe ora ricevere i dati
```
please write comments in english

prasadtalasila commented on src/main.ts:
```diff
@@ -1,84 +1,179 @@
+    });
+  }
+
+  function createTopMenu() {
```
The menu options are going to grow over a period of time. Please move this to another file.

PandoraQS replied:
The menu options have been moved in a separate file.

prasadtalasila commented on src/main.ts:
```diff
@@ -1,84 +1,179 @@
+      return;
+    }
+
+    const coeJarPath = path.join(app.getAppPath(), 'resources', 'coe', 'coe.jar');
```
Given that the application is completely dependent on Maestro, it is probably better to include maestro jar file in `resources` during the packaging of the application. Remember to use the [latest release](https://github.com/INTO-CPS-Association/maestro/releases/tag/Release%2F3.0.0).
Please add the an npm command that can execute a small javascript file that can download the latest maestro release and place it in resources. This software can be used for `npm start` and `npm test`.
In addition, please have the jar integrated into the release being built in `npm run dist` step.

prasadtalasila replied:
You can push this task to next PR if you prefer.

prasadtalasila commented on src/main.ts:
```diff
@@ -1,84 +1,179 @@
+  
+
+
+  ipcMain.handle('start-coe', async () => {
```
When the CoE is not able to start, the error is shown in the terminal but not shown to the user.

The terminal error log seems to be:

```
[1] COE JAR not found at: C:\Users\au598657\git\into-cps-application\dist\resources\coe\coe.jar
[1] Error occurred in handler for 'start-coe': Error: COE JAR not found.
[1]     at C:\Users\au598657\git\into-cps-application\dist\main.js:246:11
[1]     at WebContents.<anonymous> (node:electron/js2c/browser_init:2:86176)
[1]     at WebContents.emit (node:events:519:28)
```

PandoraQS replied:
The error message has been included:
![image](https://github.com/user-attachments/assets/9c1d046f-13b6-454a-9509-dac46b7b78a3)

prasadtalasila commented on src/resources/cosimulation/2018may7/coe.json:
`coe.json` and ``mm.json` are for a project not for the INTO-CPS Application. You can remove these two from the codebase.

PandoraQS replied:
Both `coe.json` and `mm.json` have been removed from the codebase and imported using a configuration file for the cosimulation `config.json`, which is imported using a `.env` file.
The keys imported from config.json are:
-  coeJarPath
-  simulationConfigPath
-  fmusPath
-  multiModels
-  outputPath

prasadtalasila commented on src/utils/renderJson.tsx:
Please do not render it. You can keep the main window empty for now.

PandoraQS replied:
Removed the renderJson file and the rendered .json file in the Cosimulation window.

prasadtalasila commented on src/components/Sidebar.tsx:
The application needs to be responsive. For example, please see [this code](https://github.com/INTO-CPS-Association/DTaaS/tree/feature/distributed-demo/client/src/page)

prasadtalasila reviewed (commented):
@PandoraQS Thanks for the PR. Please check the comments.

PandoraQS committed:
removed coments

PandoraQS committed:
cosimulation work in progress

PandoraQS commented:
As discussed, this is the screenshot for the cosimulation process (failing). 

![image](https://github.com/user-attachments/assets/f2909155-e74c-4b09-8ef3-b2eb6b8cefbf)

PandoraQS commented:
Simulation seems to be working now: comparing them, the result outputs.csv is the same as the one of the simulation made with the previous version of the software. Plotting outputs.csv:
![image](https://github.com/user-attachments/assets/8a8f5caf-1861-4042-bfd2-65e901a0bde5)

PandoraQS committed:
removed coe, mm and fmus from codebase, used .env for coe configuration, moved menu, fixed cosimulation and removed comments

PandoraQS committed:
Fixed sidebar collapse and resopnsiveness

prasadtalasila commented:
@PandoraQS 

Let us follow the INTO-CPS Project (this is a new project created by a user; similar to the project structure of tutorials) as proposed in [this discussion](https://github.com/INTO-CPS-Association/Co-Simulation-Studio/discussions/197).

For this PR, we can work with the following project structure:

```yaml
FMUs
  FMU1
  FMU2

cosimulation
   default/
      multi-model.json
      experiment.json

results/
  experiment_<timestamp>.csv

into-cps.yml
```

Please do update the README to reflect the working project structure expected from the user.

PandoraQS committed:
new project structure path and renamed json files in main

PandoraQS commented:
config.json (one folder above the codebase):
```
{
  "coeJarPath": "/home/simone/Scrivania/into-cps-forked_react/coe/coe.jar",
  "simulationConfigPath": "/home/simone/Scrivania/into-cps-forked_react/INTO-CPS Project/default",
  "fmusPath": "/home/simone/Scrivania/into-cps-forked_react/INTO-CPS Project/FMUS",
  "multiModels": "/home/simone/Scrivania/into-cps-forked_react/INTO-CPS Project/default/",
  "outputPath": "/home/simone/Scrivania/into-cps-forked_react/INTO-CPS Project/results"
}
```

To change the config.coe path in .env:
```
CONFIG_COE_PATH="/home/simone/Scrivania/into-cps-forked_react/config.json"
```

PandoraQS committed:
removed coe for maestro and fixed start and build configurations

SimoneMicalizzi committed:
new error for missing java, including env path for windows

PandoraQS committed:
Merge branch 'development' into cosimulation-feature-1

PandoraQS committed:
integrated maestro with existing logic for cosimulation

prasadtalasila commented on src/App.tsx:
```diff
@@ -1,51 +1,80 @@
+
+  useEffect(() => {
+    const handleResize = () => {
+      if (window.innerWidth < 768) {
```
How about an alternative. 
1. Make the sidebar responsive
2. Hide it when the sidebar width is lower than some predefined value
Please remember to define it as a constant at the top of the file and use that constant here.

PandoraQS replied:
At the current state, the Sidebar is collapsed when the width is lower than the value, but it can reopened if needed. Is it supposed to disappear completely?

prasadtalasila replied:
Collapsable option is good. Please keep it.

prasadtalasila commented on src/components/Bottom.tsx:
```diff
@@ -1,30 +1,49 @@
         bgcolor: 'background.paper',
-        boxShadow: 3,
+        boxShadow: theme.shadows[3],
+        transition: 'left 0.3s ease, width 0.3s ease',
```
what does it do? If it is required, please push the `0.3s` as a constant to top of the file

PandoraQS replied:
When left or width change, this will animate such change over 0.3 seconds with an easing function. It is mainly noticeable when the sidebar collapses, giving it a smooth animation instead of an instant change.
I will put the 0.3s as a constant

PandoraQS replied:
TRANSITION_DURATION constant has been defined.

prasadtalasila commented on src/components/Cosimulation/Cosimulation.tsx:
```diff
@@ -0,0 +1,74 @@
+  const [resultsPath, setResultsPath] = useState<string | null>(null);
+
+  useEffect(() => {
+    const handleStatusUpdate = async (event: any, status: string) => {
```
please avoid using `any`

PandoraQS replied:
I will check for every other any and avoid using it.

PandoraQS replied:
All the any statements have been removed.

prasadtalasila commented on src/components/Cosimulation/Cosimulation.tsx:
```diff
@@ -0,0 +1,74 @@
+    const handleStatusUpdate = async (event: any, status: string) => {
+      setSimulationStatus(status);
+
+      if (status === 'Simulation completed.') {
```
Are there other kinds of status? Perhaps **error**? Is it possible to use enum here?

PandoraQS replied:
Yes I think it is reasonable to enum the different statuses.

prasadtalasila commented on src/components/Cosimulation/Cosimulation.tsx:
please move the javascript code to `Cosimulation.ts`. It seems like you are following camel case for naming covention. Please change the name of this file to `CoSimulation.tsx`.

PandoraQS replied:
All the functions logic has been moved to CoSimulation.ts

prasadtalasila commented on src/components/ErrorSnackbar.tsx:
```diff
@@ -0,0 +1,34 @@
+  const [errorMessage, setErrorMessage] = useState<string>('');
+
+  useEffect(() => {
+    if (window.electronAPI) {
```
Does it make sense to split the code into the following structure?

1. Electron API and IPC is accessed in `src/electron` module. You can create an interface as required by the application here and hide all the electron interactions there.
2. Add reducers and events of react-redux store to push status events from `src/electron` module.
3. Use the redux events in the ts code being invoked from tsx code.

This structure has been used very well in [DTaaS](https://github.com/INTO-CPS-Association/DTaaS) project. Perhaps @VanessaScherma can explain this design better than I can.

PandoraQS replied:
I have tried for many hours to make Redux work with IPC interactions, but it still did not work: Redux was able to update the store of the Error statuses (needed for the snackbar), but the component did not receive any update of such store, mostly because of the different natures of the IPC interactions and Redux state management.

I find it more reasonable to centralize IPC interactions in dedicated modules (e.g., `src/electron/electronApi.ts` and `src/cosimulation/cosimulationApi.ts`) and update the UI directly through events or callbacks, bypassing Redux for these interactions.
Redux could still then be reserved for UI-related state that doesn’t depend on Electron.

Is there something missing, or maybe a way to achieve communication between both? I also have tried to use different libraries, but most of them are outdated/deprecated or they do not have any support anymore.

prasadtalasila commented on src/components/Sidebar.tsx:
```diff
@@ -1,20 +1,85 @@
+import SettingsIcon from '@mui/icons-material/Settings';
+import { NavLink } from 'react-router-dom';
+
+const drawerWidth = 240;
```
please use `MAX_OVERFLOW` like names for constants

PandoraQS replied:
The constant name has been capitalised

prasadtalasila commented on .env:
since maestro is being packaged with the application, is this needed? Please remember to use specific filename for Maestro as given on the [releases page](https://github.com/INTO-CPS-Association/maestro/releases/tag/Release%2F3.0.0).

prasadtalasila replied:
The `yarn install` command can download Maestro as well.

prasadtalasila replied:
I realise now that you are using npm. Even npm supports [pre and post](https://docs.npmjs.com/cli/v10/using-npm/scripts#pre--post-scripts) scripts. The postinstall script can be used to download Maestro.

PandoraQS replied:
The `postinstall` command has been included and maestro gets downloaded after the execution of the `npm install` command

PandoraQS replied:
However, the .env file is still required, as it loads the configuration file needed for the cosimulation (paths for multi-models, FMUS, etc.), which are still hardcoded.
An option would be to ask for the user to select the path for his personal config file, saving his path and using it to load the configuration file.

prasadtalasila commented on preload.js:
This can go into `src/electron` module

prasadtalasila replied:
If this file is a strict requirement, it might benefit from refactoring into electron and cosimulation specific parts.

PandoraQS replied:
preload.js is not strictly required for using IPC, but it is highly recommended for security and maintainability reasons.
Enabling nodeIntegration can expose the application to different vulnerabilities. 
That's the reason to use preload.js: to expose only the needed APIs and functionality to the renderer process, isolating the Node evironment.

The preload.js file has been, however, refactored, by separating it into electron and cosimulation specific parts.

prasadtalasila commented on src/resources/maestro/maestro.jar:
Please rename this to the name given on the [releases page](https://github.com/INTO-CPS-Association/maestro/releases/tag/Release%2F3.0.0)

prasadtalasila replied:
A related question. Is it necessary to put it in the codebase? Why can't something like `yarn post-install` stage download this jar file?

PandoraQS replied:
I haven't tried using `yarn post-install` as in the project there was no usage of yarn before. 
At the current state, the .jar file in the codebase was needed to package it with the application. If Maestro can be installed with yarn, I will try to do so, as it could also avoid the rabbit-hole with the temporary folder of the .jar

prasadtalasila replied:
please see the comment above regarding npm pre and post scripts.

PandoraQS replied:
The postinstall command has been included and maestro gets downloaded after the execution of the npm install comman

prasadtalasila commented on src/utils/config.ts:
```diff
@@ -0,0 +1,31 @@
+
+dotenv.config();
+
+const configFilePath = process.env.CONFIG_COE_PATH || path.resolve('./config.json');
```
This is always in the `resources/` directory of the bundled application in executable form. Why is it needed in the config file?

PandoraQS replied:
This loads the configuration file needed to run the cosimulation, not maestro itself: without configuring a path, the remaining option would be asking the user for the path.

prasadtalasila commented on test/e2e/App.spec.ts:
```diff
@@ -23,11 +23,10 @@ test.describe("General App Tests", () => {
 
         const title = await helper.window.title();
-        console.log("Window title:", title); // Log per verificare il titolo
+        console.log("Window title:", title);
```
please do not use `console.log` unless absolutely necessary. You can use during debugging but remove it again after the debugging need is over.

prasadtalasila commented on test/e2e/CoE.spec.ts:
please rename CoE to maestro. Also move all maestro-specific code to `src/maestro` and `test/maestro`

prasadtalasila commented on src/main/menu.ts:
Given that menu clicks take us to different routes, does this need to align with `react-router` part?

PandoraQS replied:
This file is used for triggering native application actions (e.g., toggling dark mode, opening developer tools, or running the cosimulation). 
At the current state of the application, there is no routing in this file.

prasadtalasila commented on src/main.ts:
please refactor electron and maestro code out of this file.

PandoraQS replied:
It has been done, but many IPC functions have to be on the main renderer.

prasadtalasila commented on src/main.ts:
```diff
@@ -1,74 +1,104 @@
+
+const maestroJarPath = isDev
+  ? path.resolve(__dirname, 'resources/maestro/maestro.jar')
+  : path.resolve(app.getAppPath(), 'dist/resources/maestro/maestro.jar');
```
Name of the maestro file can come as constant from `config.ts`

PandoraQS replied:
Name of the maestro file has been declared in `config.ts`.

prasadtalasila commented on src/main.ts:
```diff
@@ -1,74 +1,104 @@
+  }
+}
+
+function checkJavaInstallation(): Promise<boolean> {
```
This is a rabbit hole. Even if the java is installed, the java class paths could be wrong. I am thinking aloud here. Is it sufficient to start maestro and then check if it exits for any reason. The reason could be used instead of checking the underlying problems.
Again, there might be value in providing this hint to the user. You please decide either way.

PandoraQS replied:
I’ve adjusted the approach to focus on monitoring Maestro's behavior during its startup process rather than directly validating Java installation or classpath configurations, obderving Maestro's standard output and error streams to identify issues dynamically.

prasadtalasila commented on src/cosimulation/simulationContext.ts:
perhaps can be moved to `src/maestro`

PandoraQS replied:
I chose to keep the simulationContext.ts in `src/cosimulation`, which results being in the same folder of `maestro.ts`, making it reasonable to be in the same folder.  I found keeping the context in a separated file clearer, but if needed, I can include in `maestro.ts`.

prasadtalasila commented on src/cosimulation/coeApi.ts:
I like this convention. The `src/cosimulation` is better than `src/maestro`. The `src/cosimulation/maestro.ts` could still be there for interacting with Maestro at low-level.

prasadtalasila replied:
Please check [this file](https://github.com/INTO-CPS-Association/Co-Simulation-Studio/blob/cosimstudio_rework/src/maestro.ts) as well.

prasadtalasila reviewed (commented):
@PandoraQS thanks for the updates. Please check the comments.

PandoraQS committed:
constants redefinition, any statements removal, separation of cosimulation logic in different file,

PandoraQS committed:
defined a constants management structure

PandoraQS committed:
Review work in progress

PandoraQS committed:
PR #223 work in progress

PandoraQS commented:
PR Activities Review:
- Refactored MaestroManager:
  - Brought status messages together into a single file `constants/statuses.ts`.
  - Improved error handling for Java misconfiguration and startup issues with `startMaestro`, making messages more about Maestro's runtime behaviour rather than Java.
  - Ensured that `startMaestro` waits for the server to be ready before proceeding with cosimulation.
  - Improve handling of maestro JAR extraction
  - Fix port conflicts.

- Dynamic Cosimulation menu (Electron) enable/disable according to Maestro state

Added the useCosimulation hook:
 - Unifying simulation status and error state in `useCosimulation.ts`.
 - Listened for IPC messages that return simulation update, simulation error, simulation reset.

Cosimulation component refactored:
 - Logics `useCosimulation.ts` separated from the UI `CoSimulation.tsx`.
 - Dynamic showing of simulation results and error (Snackbar).

Improving the communication via IPC.
- Centralized electronAPI and cosimulationAPI in preload.ts for secure and consistent usage, while separating them for a better logical management.
  - Improved type safety with updates to global.d.ts.

- Added Maven JAR download:
  - Add postinstall script which dynamically downloads Maestro JAR from Maven repository.

- Fixed case sensitivity:
  - Standardised imports and filenames CoSimulation and useCosimulation.

- Various improvements:
  - Better TypeScript configuration with more strict checks.
  - Simulation errors are now displayed with visual feedback using Snackbar.
 
 TO BE DONE:
 - maestro.ts needs to be reviewed for a better management of the different functions.

PandoraQS committed:
Documentation for the application modules interactions, first version (electron)

PandoraQS committed:
fixed markdown lint

PandoraQS committed:
Maestro model and Maestro Model with IPC and UI

PandoraQS commented:
I have included the diagrams for the Maestro Model and the interaction between Maestro and the UI via IPC.

PandoraQS committed:
React App diagrams

PandoraQS committed:
Package diagram + react ipc communication + constant fix in app

PandoraQS committed:
fixed mardkown + wip maestro single channel configuration

PandoraQS commented:
Should I already manage the different `multi-model-X` folders, or should I just manage the `default/` folder (both for `cosimulation/` and `results/` folders)?

[Project Structure #197](https://github.com/INTO-CPS-Association/Co-Simulation-Studio/discussions/197#discussioncomment-11371887)

```
FMUs
  FMU1
  FMU2
  ...

cosimulation (one directory for each cosimulation scenario; this directory is a replacement for multi-model)
   default/ (plain co-simulation config with no concern for separate multi-models and experiment configurations
      cosim1.json
      cosim2.json
      ...
   multi-model-1/
      multi-model.json
      experiment-1
          experiment-1.json
      experiment-2/
          experiment-1.json
      ...
   multi-model-2/
      multi-model.json
      experiment-1
          experiment-1.json
      experiment-2/
          experiment-1.json
      ...

results/ (directory for storing simulation results and temporary files)
  cosimulation (same structure as top-level cosimulation; also remember to copy config files in the interest of traceability)

cosim-studio.yml (or) into-cps.yml
```

PandoraQS committed:
new project structure + new open project menu button + new maestro channels - corrected

PandoraQS committed:
unit + e2e tests + bug fixes + generic notifications handler + error handler fixes + better coe management + new statuses constants +  code cleaning

prasadtalasila commented:
TODO:

1. Documentation and demo video
2. Update description of this PR

PandoraQS committed:
Updated Application Layers Documentation and new User Guide

PandoraQS committed:
included cosimulation instructions in Main and Cosimulation pages

PandoraQS commented:
- Updated the documentation files
- Included User Guide instructions in Main and Cosimulation components

<img src="https://github.com/user-attachments/assets/3d316b7f-478d-4e2a-bc1e-a08780f2f1e9" alt="Main Page" width="420" height="350">
<img src="https://github.com/user-attachments/assets/6bddd773-f3a9-4b79-9933-d60901878f18" alt="Cosimulation Page" width="420" height="350">

PandoraQS committed:
removed user guide link from cosimulation instructions
